uniffi_bindgen/src/lib.rs

```diff
@@ -115,6 +115,133 @@ use bindings::TargetLanguage;
+/// with a type that implements [`serde::Deserialize`]. This trait represents what gets parsed from the
+/// uniffi.toml
+///
+/// # Example
+/// If the type that implements BindingGeneratorConfig is
+/// ```rs
+/// struct Config {
+///     gecko: GeckoConfig
+/// }
+/// ```
+/// Then, in the `uniffi.toml` the inner `gecko` is accessible using the toml key `gecko`, for example:
+/// ```toml
+/// [gecko]
+/// ```
+pub trait BindingGeneratorConfig:
+    for<'de> Deserialize<'de>
+    + MergeWith
+    + for<'a> From<&'a ComponentInterface>
+    + Default
+    + std::fmt::Debug
+{
+    fn load<P: AsRef<Path>>(
+        ci: &ComponentInterface,
+        crate_root: P,
+        config_file_override: Option<P>,
+    ) -> anyhow::Result<Self> {
+        let config: Self = ci.into();
+        let config_file: Option<PathBuf> = match config_file_override {
+            Some(cfg) => Some(PathBuf::from(cfg.as_ref())),
+            None => crate_root.as_ref().join("uniffi.toml").canonicalize().ok(),
+        };
+        match config_file {
+            Some(path) => {
+                let contents = slurp_file(&path)
+                    .with_context(|| format!("Failed to read config file from {:?}", &path))?;
+                let loaded_config: Self = toml::de::from_str(&contents)
+                    .with_context(|| format!("Failed to generate config from file {:?}", &path))?;
+                Ok(loaded_config.merge_with(&config))
+            }
+            None => Ok(config),
+        }
+    }
+}
+
+/// A trait representing a UniFFI Binding Generator
+///
+/// External crates that implement binding generators, should implement this type
+/// and call the [`generate_external_bindings`] using a type that implements this trait.
+pub trait BindingGenerator: Sized {
+    /// Associated type representing a the configuration parsed from the
+    /// uniffi.toml
+    type Config: BindingGeneratorConfig;
+
+    /// A Constructor, allows passing configuration that was
+    /// parsed from the `uniffi.toml` to the type that implements
+    /// BindingGenerator
+    fn new(config: Self::Config) -> Self;
+
+    /// Writes the bindings to the output directory
+    ///
+    /// # Arguments
+    /// - `ci`: A reference to a [`ComponentInterface`] representing the interface
+    /// - `out_dir`: The path to where the binding generator should write the output bindings
+    fn write_bindings<P: AsRef<Path>>(
+        &self,
+        ci: &ComponentInterface,
+        out_dir: P,
+    ) -> anyhow::Result<()>;
+
+    /// Compiles the bindings that are written by `write_bindings`, this is only relevant to run tests
+    /// and for languages that are compiled
+    ///
+    /// # Arguments
+    /// - `ci`: A reference to a [`ComponentInterface`] representing the interface
+    /// - `out_dir`: The path to where the binding generator should write the output bindings
+    fn compile_bindings<P: AsRef<Path>>(
```

Comment:
I'm not sure if we want to remove `compile_bindings` and `run_script` as a part of this, so I just gave them default implementations. Right now, they are not called anywhere. I see two possible futures here:
1. We remove them from the trait, and have the binding generator handle how testing works
1. Create another generic function, similar to `generate_external_bindings` that does the testing - similar to how `run_tests` works now